src/array.rs

```diff
@@ -0,0 +1,456 @@
+use std::ffi::c_void;
+
+use half::{bf16, f16};
+use mlx_sys::mlx_array;
+use num_complex::Complex;
+
+use crate::sealed::Sealed;
+
+// TODO: camel case?
+// Not using Complex64 because `num_complex::Complex64` is actually Complex<f64>
+#[allow(non_camel_case_types)]
+pub type c64 = Complex<f32>;
+
+/// Array element type
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

Comment:
for copy, we'd want to make this happen:

```rust
mlx_retain(ctx)
return Array:: from_ptr(ctx)
```

Reply:
That was a mistake. Didn't pay attention to what copilot generated fo the derive. I think we should actually forbid `Clong` or `Copy`. `mlx_retain` just increment the counter but the buffer is not guarded by a `Mutex` or anything similar

Reply:
For Copy this should be ok though right? the counter is incremented and we track the same space in memory?

Reply:
The mlx_array is essentially `Arc<Bytes>`. Maybe wait until we see how many ops mutate the array? If `eval` is the only one that mutate the array, `Clone` and `Copy` is probably fine, because the c++ code checks if the array has been evaluated (if I remember correctly)

Reply:
Another potential problem with `Clone` and `Copy`. There is a lot of `std::move` in the ops c++ code, not sure if the underlying `mlx_array` ptr is still valid after those ops

Reply:
Just realized that this is actually `Clone` and `Copy` on `Dtype`. Let's move the conversation to `Array`